controller/konnect/ops/ops_kongvault.go

```diff
@@ -0,0 +1,158 @@
+package ops
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"slices"
+
+	sdkkonnectcomp "github.com/Kong/sdk-konnect-go/models/components"
+	sdkkonnectops "github.com/Kong/sdk-konnect-go/models/operations"
+	sdkkonnecterrs "github.com/Kong/sdk-konnect-go/models/sdkerrors"
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+	"github.com/kong/kubernetes-configuration/pkg/metadata"
+)
+
+func createVault(ctx context.Context, sdk VaultSDK, vault *configurationv1alpha1.KongVault) error {
+	cpID := vault.GetControlPlaneID()
+	if cpID == "" {
+		return fmt.Errorf(
+			"can't create %T %s without a Konnect ControlPlane ID",
+			vault, client.ObjectKeyFromObject(vault),
+		)
+	}
+
+	vaultInput, err := kongVaultToVaultInput(vault)
+	if err != nil {
+		return fmt.Errorf("failed to convert KongVault to Konnect vault input: %w", err)
+	}
+	resp, err := sdk.CreateVault(ctx, cpID, vaultInput)
+
+	if errWrapped := wrapErrIfKonnectOpFailed(err, CreateOp, vault); errWrapped != nil {
+		SetKonnectEntityProgrammedConditionFalse(vault, "FailedToCreate", errWrapped.Error())
+		return errWrapped
+	}
+
+	vault.SetKonnectID(*resp.Vault.ID)
+	SetKonnectEntityProgrammedCondition(vault)
+	return nil
+}
+
+func updateVault(ctx context.Context, sdk VaultSDK, vault *configurationv1alpha1.KongVault) error {
+	cpID := vault.GetControlPlaneID()
+	if cpID == "" {
+		return fmt.Errorf(
+			"can't update %T %s without a Konnect ControlPlane ID",
+			vault, client.ObjectKeyFromObject(vault),
+		)
+	}
+	id := vault.GetKonnectID()
+	vaultInput, err := kongVaultToVaultInput(vault)
+	if err != nil {
+		return fmt.Errorf("failed to convert KongVault to Konnect vault input: %w", err)
+	}
+
+	_, err = sdk.UpsertVault(ctx, sdkkonnectops.UpsertVaultRequest{
+		VaultID:        id,
+		ControlPlaneID: cpID,
+		Vault:          vaultInput,
+	})
+
+	if errWrapped := wrapErrIfKonnectOpFailed(err, CreateOp, vault); errWrapped != nil {
+		// Service update operation returns an SDKError instead of a NotFoundError.
+		var sdkError *sdkkonnecterrs.SDKError
+		if errors.As(errWrapped, &sdkError) {
+			switch sdkError.StatusCode {
+			// REVIEW: should we use constants defined in `net/http` or numerics here?
```

Comment:
I believe either is fine. Given that other ops files use numerics I'd stay consistent and use that in here as well.

Reply:
Done.